.circleci/config.yml

```diff
@@ -178,18 +178,39 @@ jobs:
       - store_test_output
       - fail_fast
       - teardown_vm
-  'Unit Tests':
+  'Local Unit Tests':
     executor:
       name: amphtml-large-executor
     steps:
+      - unless:
+          condition:
+            matches:
+              pattern: ^pull/\d+$
+              value: << pipeline.git.branch >>
+          steps:
+            - run:
+                name: 'Skip Local Unit Test on Push Builds'
+                command: circleci-agent step halt
       - setup_vm
       - install_chrome
       - run:
-          name: 'Unit Tests'
-          command: node build-system/pr-check/unit-tests.js
+          name: 'Local Unit Tests'
+          command: node build-system/pr-check/unit-tests-local.js
       - store_test_output
       - fail_fast
       - teardown_vm
+  'All Unit Tests':
+    executor:
+      name: amphtml-large-executor
+    parallelism: 16
```

Comment:
CircleCI gives our entire org a max of ~80 parallel VMs, after which it starts queueing. Do we know if this number counts towards that quota? (I'm wondering if this PR could result in more queueing once it is merged, even if the builds run faster.)

Can we experiment with lower parallelism numbers (say, 4 and 8) and see what the effects are on overall CPU consumption and running time? (E.g. each parallel shard appears to be doing things like checking out code, installing packages, installing chrome, etc.)

I'm guessing the sweet spot lies below 16, but it would be nice to back that up with real results, and reduce any potential cause for increased queueing of jobs.

Reply:
I experimented with multiple values, 16 seems to be the sweet spot - above 16 the improvement in runtime is marginal-to-nonexistent

As for the shared installation steps, I've seen something in the CircleCI documentation that makes me think we can speed those up (not 100% sure, but definitely want to experiment with it soon)

Reply:
It's a bit difficult to gather because I didn't keep the data around, but here are some samples:

- End-to-End tests with 8 shards: [5m27s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12938/workflows/0c0118ef-9d3d-4ac7-8b26-2adf17bab69d/jobs/207377)
- Unit Tests with 4 shards: [more than 8m55s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12897/workflows/307caaa1-02e6-4af3-9cb5-8e1f785c2ff1/jobs/206565) ("more" because `fail_fast` stopped these before they completed)
- Unit Tests with 8 shards: [more than 3m35s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12926/workflows/ff311539-6fc2-4d41-9f76-1677d727b596/jobs/207111)
- Unit Tests with 32 shards: [4m31s](https://app.circleci.com/pipelines/github/ampproject/amphtml/12928/workflows/20a9e7fb-40b6-4e0d-a0e2-e2c7c9f229ab/jobs/207161)

With 16 shards for each, they take between 3m and 3m30s. Admittedly this isn't super scientific, but I think this is Good Enough™ for now

Reply:
I compared recent [`main` branch builds](https://app.circleci.com/insights/github/ampproject/amphtml/workflows/CircleCI/overview?branch=main&reporting-window=last-7-days) against [builds for this PR branch](https://app.circleci.com/insights/github/ampproject/amphtml/workflows/CircleCI/overview?branch=pull%2F35227&reporting-window=last-7-days):

`main`: ~2200 credits for ~16 minute builds.

![image](https://user-images.githubusercontent.com/26553114/126390683-37a6e4f0-820e-47ea-8368-0e19b902996f.png)


This PR: ~4300 credits for ~14 minute builds.

![image](https://user-images.githubusercontent.com/26553114/126390662-849c0318-d061-41bb-ac0e-56ebe7fc0f95.png)

By increasing parallelism to 16, it appears we're spending nearly 100% more $$ for a < 15% speed up.
- Do these numbers seem right to you?
- Can we retry with parallelism factors of 4 and 8 to see if there's a better sweet spot for total duration vs. total credit usage for an average build (that affects the AMP runtime)?

Reply:
Ah! So they _do_ charge extra for the parallelization. ok, let's try a few executions with 8 / 4 parallels

Reply:
Yep. And while you're at it, can we generate a build with 2 parallel shards as well? After that, once they all show up on the insights dashboard for this PR, we can compare costs and benefits for 2 through 16. (These numbers don't have to all be powers of 2 😃 )

Reply:
I experimented with parallelism numbers and Docker container sizes for the newly-parallelized jobs, and collected the data here: https://docs.google.com/spreadsheets/d/11JOAxtNTCzV7KxKXAwVrHhjHvsoVBW-JKEmsTL8BK3U/edit#gid=1450490992

Given the averages, setting parallelism to 6 and container size to medium seems like a good choice - it increases the credits used from (average of) 2,138 to 2,388 but reduces the (average) total duration from 18:09 minutes to 13:58 minutes, which is a huge improvement at a marginal cost.

With followup PRs I believe I can lower the credits used again

Reply:
This is brilliant! Thanks for going through this exercise. I agree with your conclusion.